src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,173 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            linkFirst(value);
+        } else {
+            linkLast(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexInBoundsStrict(index);
+        if (index == 0) {
+            linkFirst(value);
+        } else if (index > 0 && index == size) {
+            linkLast(value);
+        } else {
+            link(value, index);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInBounds(index);
+        return find(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = find(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = find(index);
+        T oldValue = node.value;
+        unlink(node);
+        size--;
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> node = head; node != null; node = node.next) {
+            if (object == node.value
+                    || object != null && object.equals(node.value)) {
+                unlink(node);
+                size--;
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexInBounds(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(outOfBoundsMessage() + index);
```

Comment:
you can change the signature of `outOfBoundsMessage()` to `outOfBoundsMessage(int index)` to avoid concatenation